es6/animations.js

```diff
@@ -5,180 +5,176 @@ import {
+            && !(val1 === undefined && val2 === undefined))
+}
 
-        if (key !== undefined)
-            return checked_animation_amt({start_state: start_state[key], end_state: end_state[key], amt: amt[key]})
+const isNumber = (val) => {
+    return typeof(val) === 'number'
+}
 
-        if (typeof(start_state) !== 'object' || typeof(end_state) !== 'object' || typeof(amt) !== 'object') {
-            throw 'Incompatible types passed as {start_state, end_state, amt}, must all be objects or numbers'
-        }
+const validateAnimation = (animation) => {
+    const end_time = animation.end_time
+    const end_state = animation.end_state
 
-        let keys = Object.keys(start_state)
-        if (!keys.length) keys = Object.keys(end_state)
-        if (!keys.length) keys = Object.keys(amt)
-
-        return keys.reduce((acc, key) => {
-            const amts = checked_animation_amt({start_state: start_state[key], end_state: end_state[key], amt: amt[key]})
-            acc.start_state[key] = amts.start_state
-            acc.end_state[key] = amts.end_state
-            acc.amt[key] = amts.amt
-            return acc
-        }, {start_state: {}, end_state: {}, amt: {}})
+    const computed_end_state = computeAnimatedState([animation], _end_time)
+    if (!isEqual(computed_end_state, end_state)) {
+        throw `Invalid Animate: end_state !== computed_end_state for animation:`
+            + `\n${JSON.stringify(animation, null, '  ')}:`
+            + `${JSON.stringify(computed_end_state, null, '  ')} !==`
+            + `${JSON.stringify(end_state, null, '  ')}`
     }
 }
 
-const KeyedAnimation = ({type, path, key, start_time, end_time, duration, start_state, end_state, amt, curve, unit}) =>
-    Animate({
-        type,
-        path: `${path}/${key}`,
-        key,
-        start_time,
-        end_time,
-        duration,
-        start_state: start_state && start_state[key],
-        end_state: end_state && end_state[key],
-        amt: amt && amt[key],
-        curve, unit,
-    })
+export const Animate = ({
+        type, path, start_time, end_time, duration, start_state, end_state,
+        delta_state, merge=false, curve='linear', unit=null, tick=null,
+        error_checking=true}) => {
 
+    let _start_time, _end_time, _duration,
+        _end_state, _delta_state, _tick, _split_path, _type
 
-export const Become = ({path, state, start_time, end_time=Infinity, duration=Infinity}) => {
-    if (start_time === undefined) start_time = (new Date).getTime()
+    _start_time = (start_time === undefined) ? Date.now() : start_time
 
-    var {start_time, end_time, duration} = checked_animation_duration({start_time, end_time, duration})
-    if (start_time === undefined || path === undefined) {
-        console.log({path, state, start_time, end_time, duration})
-        throw 'Become animation must have a start_time and path defined.'
+    if (exactlyOneIsUndefined(delta_state, end_state)) {
```

Comment:
Dunno, this seems way cleaner to me:

```javascript
animation = {...animation, computeDeltaState({start_state, delta_state, end_state)}
animation = {...animation, computeDeltaTime({start_time, duration, end_time)}
```

`computeDeltaState` then does both the type and diff checks.